nRF Desktop Failsafe module: migration to the Zephyr HWinfo driver and activation on the nRF54H20 DK

kapi-no opened the pull request:
Depends on:

~https://github.com/nrfconnect/sdk-nrf/pull/19028~

Issues:
- ~Watchdog reset reason bit is always set for the nRF54H20 DK regardless whether the reset is caused by the reset pin (`RESET_PIN`: `0x00000011`) or software reset (`RESET_SOFTWARE`: `0x00000012`).~

Out standing items:
- [x] Changelog and the documentation update for the Failsafe module.

NordicBuilder commented:
The following west manifest projects have been modified in this Pull Request:

| Name | Old Revision | New Revision | Diff |
| ---- | ------------ | ------------ |------|
| zephyr | https://github.com/nrfconnect/sdk-zephyr/commit/8bc628d8945cb7bb01f57c92c00b5f1550ed576c | https://github.com/nrfconnect/sdk-zephyr/pull/2293 | [nrfconnect/sdk-zephyr#2293/files](https://github.com/nrfconnect/sdk-zephyr/pull/2293/files) |

*Note: This message is automatically posted and updated by the Manifest GitHub Action.*

NordicBuilder commented:
## CI Information

<sup>_To view the history of this post, clich the 'edited' button above_</sup>
Build number: 1
### Inputs:
#### Sources:
<details><summary> more details</summary>
</details>

#### Github labels
| Enabled | Name | Description |
| -- | -- | -- |
|| ci-disabled | _Disable the ci execution_ | 
|| ci-all-test | _Run all of ci, no test spec filtering will be done_ | 
|| ci-force-downstream | _Force execution of downstream even if twister fails_ | 
|| ci-run-twister | _Force run twister_ | 
|| ci-run-zephyr-twister | _Force run zephyr twister_ | 

<details><summary> List of changed files detected by CI (0)</summary>

```
```
</details>

### Outputs:

#### Toolchain
Version: 
Build docker image: 

#### Test Spec & Results: :white_check_mark: Success; :x: Failure; :orange_circle: Queued; :yellow_circle: Progress; :white_medium_square: Skipped; :warning: Quarantine
- :x: Toolchain
- :x: Build twister
- :x: Integration tests



*Note: This message is automatically posted and updated by the CI*

MarekPieta commented on applications/nrf_desktop/src/modules/failsafe.c:
```diff
@@ -15,14 +15,24 @@
+
+	*ret = ((reas & mask) != 0);
+
+	LOG_INF("Reset reason (0x%08X) trigger %s for Failsafe erase operation",
```
Can we make it a bit shorter: `"Reset reason (0x%08X), trigger %s`? Log would be prefixed with module name by the logger anyway

MarekPieta commented on applications/nrf_desktop/src/modules/failsafe.c:
```diff
@@ -43,9 +53,11 @@ static void failsafe_erase(void)
+static int failsafe_clear(void)
 {
-	nrfx_reset_reason_clear(nrfx_reset_reason_get());
+	int err = hwinfo_clear_reset_cause();
```
`return hwinfo_clear_reset_cause();` - for simplicity?

kapi-no replied:
I added some additional error logs

MarekPieta commented on applications/nrf_desktop/src/modules/failsafe.c:
```diff
@@ -55,17 +67,27 @@ static bool app_event_handler(const struct app_event_header *aeh)
 
-			module_set_state(MODULE_STATE_READY);
+			if (!err) {
+				err = failsafe_clear();
```
Consider avoiding clearing the reset reason if `failsafe_erase` call fails.

kapi-no replied:
This worked a bit differently in the previous implementation, where we cleared the reset reason unconditionally. Are you sure about that?

MarekPieta replied:
I am aware that it used to work differently.
I am wondering a bit which approach would be better here: if we desist from clearing the reset reason on settings erase error, we give failsafe module another chance to erase settings which could be safer.

kapi-no commented on applications/nrf_desktop/src/modules/failsafe.c:
```diff
@@ -15,14 +15,24 @@
 
 
-static bool failsafe_check(void)
+static bool failsafe_check(bool *ret)
```
I prefer failure detected as the settings erase may be one of many actions in the future that this module needs to perform on failure detection

kapi-no commented on applications/nrf_desktop/src/modules/failsafe.c:
```diff
@@ -55,17 +67,27 @@ static bool app_event_handler(const struct app_event_header *aeh)
 
 		if (check_state(event, MODULE_ID(main), MODULE_STATE_READY)) {
+			int err;
+			bool check_result;
```
`failure_detected` for consistency

kapi-no commented:
@MarekPieta, addressed your comments

kapi-no commented:
The Watchdog issue should be fixed after addressing the comments from the Zephyr upstream PR that adds reset reason support for nRF54H20 SoC in the Zephyr HWinfo driver:

https://github.com/zephyrproject-rtos/zephyr/pull/81751

MarekPieta commented on applications/nrf_desktop/src/modules/failsafe.c:
```diff
@@ -15,17 +15,28 @@
 
 
-static bool failsafe_check(void)
+static bool failsafe_check(bool *failure_detected)
```
`static int`

kapi-no replied:
thanks, fixed

kapi-no commented:
The issue is resolved as the HW info driver fixes are merged as part of the following PR:

https://github.com/nrfconnect/sdk-zephyr/pull/2385

NordicBuilder commented:
You can find the documentation preview for this PR at this <a href='https://ncsdoc.z6.web.core.windows.net/PR-19142
/'>link</a>. It will be updated about 10 minutes after the documentation build succeeds.

*Note: This comment is automatically posted by the Documentation Publish GitHub Action.*

divipillai commented on doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:
```diff
@@ -198,7 +198,11 @@ nRF5340 Audio
+* Updated:
+
+  * The :ref:`nrf_desktop_failsafe` to use the Zephyr :ref:`zephyr:hwinfo_api` driver for getting and clearing the reset reason information (see the :c:func:`hwinfo_get_reset_cause` and :c:func:`hwinfo_clear_reset_cause` function).
+    The Zephyr :ref:`zephyr:hwinfo_api` driver replaces the nrfx reset reason helper depedency (see the :c:func:`nrfx_reset_reason_get` and :c:func:`nrfx_reset_reason_clear` function).
```
```suggestion
    The Zephyr :ref:`zephyr:hwinfo_api` driver replaces the nrfx reset reason helper dependency (see the :c:func:`nrfx_reset_reason_get` and :c:func:`nrfx_reset_reason_clear` functions).
```

divipillai commented on doc/nrf/releases_and_maturity/releases/release-notes-changelog.rst:
```diff
@@ -198,7 +198,11 @@ nRF5340 Audio
-|no_changes_yet_note|
+* Updated:
+
+  * The :ref:`nrf_desktop_failsafe` to use the Zephyr :ref:`zephyr:hwinfo_api` driver for getting and clearing the reset reason information (see the :c:func:`hwinfo_get_reset_cause` and :c:func:`hwinfo_clear_reset_cause` function).
```
```suggestion
  * The :ref:`nrf_desktop_failsafe` to use the Zephyr :ref:`zephyr:hwinfo_api` driver for getting and clearing the reset reason information (see the :c:func:`hwinfo_get_reset_cause` and :c:func:`hwinfo_clear_reset_cause` functions).
```

kapi-no commented on applications/nrf_desktop/src/modules/failsafe.c:
```diff
@@ -15,17 +15,28 @@
-			NRFX_RESET_REASON_LOCKUP_MASK;
+	int err;
+	uint32_t reas = 0;
+	uint32_t mask = RESET_WATCHDOG |
```
improved

kapi-no commented:
Rebased due to the CI failures. No file diff has changed (pure rebase).

kapi-no commented:
@nordicjm, could you review the build system part?

kapi-no commented:
Rebased due to the CI failures. No file diff has changed (pure rebase).
